lib/storage/providers/azure.rb

```diff
@@ -0,0 +1,294 @@
+                        .map(&:name)
+                        .sort
+
+        (dirs + files).join("\n")
+      end
+    end
+
+    private
+
+    def split_path(src)
+      path = src.split('/')
+      dir = path.length == 1 ? '' : path[..-2].join('/')
+
+      [dir, path.last]
+    end
+
+    def get_file_properties(src)
+      dir, file = split_path(src)
+
+      client.get_file_properties(
+        file_share_name,
+        dir,
+        file
+      ).properties
+    end
+
+    def delete_file(src)
+      dir, file = split_path(src)
+
+      # ensure directory exists, as client doesn't tell you and
+      # continues as normal if the directory doesn't exist
+      query_directory(dir)
+
+      client.delete_file(
+        file_share_name,
+        dir,
+        file
+      )
+
+      src
+    rescue Azure::Core::Http::HTTPError => e
+      if e.message.include?("resource does not exist")
+        raise ResourceNotFoundError.new(src)
+      end
+    end
+
+    def get_file(src)
+      dir, file = split_path(src)
+
+      # ensure directory exists, as client doesn't tell you and
+      # continues as normal if the directory doesn't exist
+      query_directory(dir)
+
+      file, content = client.get_file(
+        file_share_name,
+        dir,
+        file
+      )
+      
+      content
+    rescue Azure::Core::Http::HTTPError => e
+      if e.message.include?("resource does not exist")
+        raise ResourceNotFoundError.new(src)
+      end
+    end
+
+    def chunk_upload(src, target_dir, target_file, filesize)
+      max_upload = 4_000_000
+      number_of_chunks = (filesize.to_f / max_upload).ceil
+
+      File.open(src) do |f|
+        number_of_chunks.times do |iter|
+          to_read = max_upload
+          buffer = f.read(to_read)
+
+          start_range = iter * max_upload
+          end_range = ((iter * max_upload) + buffer.bytesize) -1
+
+          client.put_file_range(
+            file_share_name,
+            target_dir,
+            target_file,
+            start_range,
+            end_range,
+            buffer
+          )
+        end
+      end
+    end
+
+    def chunk_download(src, dest, filesize)
+      dir, file = split_path(src)
+
+      number_of_chunks = (filesize.to_f / MAX_FILESIZE).ceil
+      chunk_files = []
+
+      number_of_chunks.times do |iter|
+        start_range = iter * MAX_FILESIZE
+        end_range = iter == number_of_chunks - 1 ?
+          filesize :
+          ((iter + 1) * MAX_FILESIZE) -1
+
+
+        options = {
+          start_range: start_range,
+          end_range: end_range
+        }
+
+        _, content = client.get_file(
+          file_share_name,
+          dir,
+          file,
+          options
+        )
+
+        tempfile = Tempfile.new("#{file}-#{iter+1}.chunk")
+        chunk_files << tempfile
+
+        tempfile.write(content)
+      end
+
+      File.open(dest, 'w+') do |f|
+        `cat #{chunk_files.map(&:path).join(' ')} > #{dest}`
+      end
```

Comment:
Is making temporary files necessary here? If you open the file before starting the loop and call `File.write(content)` every iteration it'll append it to the existing file as you're downloading chunks, and then there's no need to `require tempfile`, or save so many different tempfiles, or concatenate them all at the end (similar to the way you're uploading chunks, reading portions of the file and uploading each one after reading it).

Reply:
The main reason I did it this way is because I have a philosophy with file reading/writing which says "get in and out as quickly as possible". Maintaining a `File.open` block for the entirety of a file download means that it's open for far longer than it needs to be. It also means that the file will immediately be created, when it may or may have finished downloading. The user/another process may see a file and think that it's done when it's actually still being concatenated to.

Reply:
I see, that does make sense. The default behaviour for AWS is indeed to create an empty file and start writing to it until it's finished - perhaps I could change the AWS implementation to initially make the file hidden and rename it to something non-hidden when it's finished downloading? I'm not sure if tempfiles are visible by default but it might be worth considering that for the azure implementation as well.

Reply:
It's not that tempfiles are invisible; the Ruby `Tempfile` class creates files at `/tmp` and deletes the `Tempfile` objects either via an explicit close or when the Ruby interpreter exists ([docs](https://ruby-doc.org/stdlib-2.5.3/libdoc/tempfile/rdoc/Tempfile.html)).